Add GCP documentation (#127)
DefangLabs/defang-docs: docs/providers/gcp.md

sidebar_position: 3000
---

# Google Cloud Platform (GCP)

:::info

> The Defang GCP provider is available for Public Preview as of December 2024. This page will be updated when it is GA.

Author: ```suggestion
The Defang GCP provider is available for Public Preview as of December 2024.
```

:::

> Defang enables you to effortlessly create and manage full, scalable applications with GCP. It is designed to simplify deploying your services to the cloud. As one of the leading cloud providers globally, GCP offers powerful tools and resources, and with Defang, you can bypass the complexities and challenges of the GCP platform. Let Defang handle the heavy lifting so you can focus on what matters most to you!

Author: ```suggestion
Defang enables you to effortlessly develop and deploy full, scalable applications with GCP. It is designed to simplify deploying your services to the cloud. As one of the leading cloud providers globally, GCP offers powerful tools and resources, and with Defang, you can bypass the complexities of the GCP platform. Let Defang handle the heavy lifting so you can focus on what matters most to you!
```

> Defang enables you to effortlessly create and manage full, scalable applications with GCP. It is designed to simplify deploying your services to the cloud. As one of the leading cloud providers globally, GCP offers powerful tools and resources, and with Defang, you can bypass the complexities and challenges of the GCP platform. Let Defang handle the heavy lifting so you can focus on what matters most to you!

Reviewer: In the first sentence, recommend replace "create and manage" with "develop and deploy".
Recommend delete the second sentence.
In the 3rd sentence recommend delete "and challenges".

## Getting Started

> After signing in to your GCP account, select an existing project or [create a new project](https://developers.google.com/workspace/guides/create-project), make sure [billing is enabled](https://cloud.google.com/billing/docs/how-to/modify-project) and note down the project ID and set it as environment variable `GCP_PROJECT_ID`.

Reviewer: missing comma after (https://cloud.google.com/billing/docs/how-to/modify-project)

> After signing in to your GCP account, select an existing project or [create a new project](https://developers.google.com/workspace/guides/create-project), make sure [billing is enabled](https://cloud.google.com/billing/docs/how-to/modify-project) and note down the project ID and set it as environment variable `GCP_PROJECT_ID`.

Author: ```suggestion
After signing in to your GCP account, select an existing project or [create a new project](https://developers.google.com/workspace/guides/create-project), make sure [billing is enabled](https://cloud.google.com/billing/docs/how-to/modify-project), and note down the project ID and set it as environment variable `GCP_PROJECT_ID`.
```

```bash
export GCP_PROJECT_ID=<your-project-id>
```

> Next step is to [authenticate your local environment with GCP](https://cloud.google.com/docs/authentication), our preferred method is to setup [Application Default Credentials](https://cloud.google.com/docs/authentication/provide-credentials-adc) with the google cloud CLI. Once the [google cloud CLI is installed](https://cloud.google.com/sdk/docs/install), run the following command to authenticate:

Reviewer: Turn this comma into period and start a new sentence: ...with GCP](https://cloud.google.com/docs/authentication),
One occurrence of Google is not capitalized ("google cloud CLI").

> Next step is to [authenticate your local environment with GCP](https://cloud.google.com/docs/authentication), our preferred method is to setup [Application Default Credentials](https://cloud.google.com/docs/authentication/provide-credentials-adc) with the google cloud CLI. Once the [google cloud CLI is installed](https://cloud.google.com/sdk/docs/install), run the following command to authenticate:

Author: ```suggestion
Next step is to [authenticate your local environment with GCP](https://cloud.google.com/docs/authentication). Our preferred method is to setup [Application Default Credentials](https://cloud.google.com/docs/authentication/provide-credentials-adc) with the Google Cloud CLI. Once the [Google Cloud CLI is installed](https://cloud.google.com/sdk/docs/install), run the following command to authenticate:
```

```bash
gcloud init
gcloud auth application-default login
```

> The Defang CLI will automatically check if they are set before running. Once you are ready to go, add the `--provider=gcp` to your command to tell the Defang CLI to use the GCP provider.

Reviewer: First sentence - what does "they" refer to?

> The Defang CLI will automatically check if they are set before running. Once you are ready to go, add the `--provider=gcp` to your command to tell the Defang CLI to use the GCP provider.

Author: ```suggestion
The Defang CLI will automatically check if `GCP_PROJECT_ID` environment variable is set and correctly authenticated with GCP before running. Once you are ready to go, add the `--provider=gcp` to your command to tell the Defang CLI to use the GCP provider.
```

```bash
$ defang up --provider=gcp
# or
$ DEFANG_PROVIDER=gcp defang up
```

## Location

The Defang BYOC GCP Provider will use the location specified in the `GCP_LOCATION` environment variable. For a list of locations available in GCP, see the [location documentation](https://cloud.google.com/about/locations). If the `GCP_LOCATION` environment variable is not set, the default location `us-central1` (Iowa) will be used.

## Architecture

> Defang uses GCP cloud run to build, deploy and run your services. The following describes the current state of Defang's support for GCP, the specific resources that Defang uses, and the roadmap for future support.

Reviewer: Missing comma after deploy

> Defang uses GCP cloud run to build, deploy and run your services. The following describes the current state of Defang's support for GCP, the specific resources that Defang uses, and the roadmap for future support.

Author: ```suggestion
Defang uses GCP cloud run to build, deploy, and run your services. The following describes the current state of Defang's support for GCP, the specific resources that Defang uses, and the roadmap for future support.
```

### Deployment

> To deploy your services, the Defang CLI will setup some basic resources needed, including enabled required APIs in the project, create service accounts used to build and deploy your service with required permissions, and a [Google Cloud Storage](https://cloud.google.com/storage) bucket where Defang CLI packages your code and uploads it to. The CLI then deploys a Cloud Run Job that uses Pulumi to build your container image and run your services.

Reviewer: The first sentence is quite long and mixes tenses. Not 100% sure, but something like this?
To deploy your services, the Defang CLI sets up some basic resources needed, including enabling required APIs in the project, creating service accounts used to build and deploy your service with the required permissions, and creating a [Google Cloud Storage](https://cloud.google.com/storage) bucket where the Defang CLI uploads your source code to.

> To deploy your services, the Defang CLI will setup some basic resources needed, including enabled required APIs in the project, create service accounts used to build and deploy your service with required permissions, and a [Google Cloud Storage](https://cloud.google.com/storage) bucket where Defang CLI packages your code and uploads it to. The CLI then deploys a Cloud Run Job that uses Pulumi to build your container image and run your services.

Reviewer: "Cloud Run Job" should be "GCP Cloud Run job"?

> To deploy your services, the Defang CLI will setup some basic resources needed, including enabled required APIs in the project, create service accounts used to build and deploy your service with required permissions, and a [Google Cloud Storage](https://cloud.google.com/storage) bucket where Defang CLI packages your code and uploads it to. The CLI then deploys a Cloud Run Job that uses Pulumi to build your container image and run your services.

Author: ```suggestion
To deploy your services, the Defang CLI sets up some basic resources needed, including enabling required APIs in the project, creating service accounts used to build and deploy your service with the required permissions, and creating a [Google Cloud Storage](https://cloud.google.com/storage) bucket where the Defang CLI uploads your source code to. The CLI then deploys a GCP Cloud Run Job that uses Pulumi to build your container image and run your services.
```

### Runtime

> The Provider build and deploy your services using the [Google Cloud Run](https://cloud.google.com/run) jobs, and runs your workloads using the [Google Cloud Run](https://cloud.google.com/run) service.

Reviewer: The Provider builds and deploys your services using [Google Cloud Run](https://cloud.google.com/run) jobs, and runs your workloads using the [Google Cloud Run](https://cloud.google.com/run) service.

> The Provider build and deploy your services using the [Google Cloud Run](https://cloud.google.com/run) jobs, and runs your workloads using the [Google Cloud Run](https://cloud.google.com/run) service.

Author: ```suggestion
The Provider builds and deploys your services using [Google Cloud Run](https://cloud.google.com/run) jobs, and runs your workloads using the [Google Cloud Run](https://cloud.google.com/run) service.
```
